File: src/server/calculator.js
/**
 * Postage Calculator - Implements tariff and shipping cost calculations
 * Based on Australia Post rates, US IEEPA tariffs, and Zonos processing fees
 */

const postalRates = require('../data/postalRates.json');
const tariffs = require('../data/tariffs.json');
const brands = require('../data/brands.json');
const extraCoverData = require('../data/extraCover.json');

/**
 * Get the country of origin for a brand
 * @param {string} brandName - The brand name
 * @returns {string} - Country of origin
 */
function getCountryOfOrigin(brandName) {
  const brand = brands.brands[brandName];
  if (brand) {
    return brand.primaryCOO;
  }
  return brands.defaultCOO; // Default to China if unknown
}

/**
 * Get the tariff rate for a country
 * @param {string} country - Country of origin
 * @returns {number} - Tariff rate as decimal (e.g., 0.20 for 20%)
 */
function getTariffRate(country) {
  return tariffs.usaTariffs.rates[country] ?? tariffs.usaTariffs.rates['China'];
}

/**
 * Calculate AusPost shipping cost
 * @param {string} zone - Destination zone (e.g., "3-USA & Canada")
 * @param {string} weightBand - Weight category (XSmall, Small, Medium, Large, XLarge)
 * @param {number} discountBand - Discount band (0-5)
 * @returns {number} - Shipping cost in AUD
 */
function calculateAusPostShipping(zone, weightBand, discountBand = 0) {
  const zoneData = postalRates.zones[zone];
  if (!zoneData) {
    throw new Error(`Unknown zone: ${zone}`);
  }

  const weightData = zoneData.weightBands[weightBand];
  if (!weightData) {
    throw new Error(`Unknown weight band: ${weightBand}`);
  }

  const basePrice = weightData.basePrice;
  const handlingFee = zoneData.handlingFee;
  const discount = zoneData.discountBands[discountBand] || 0;

  // Formula: Base × (1 + handling) × (1 - discount)
  const withHandling = basePrice * (1 + handlingFee);
  const finalPrice = withHandling * (1 - discount);

  return Math.round(finalPrice * 100) / 100;
}

/**
 * Calculate extra cover (insurance) cost
 * @param {number} itemValueAUD - Item value in AUD
 * @param {number} discountBand - Discount band (0-5)
 * @returns {number} - Extra cover cost in AUD
 */
function calculateExtraCover(itemValueAUD, discountBand = 0) {
  const { basePricePer100, thresholdAUD, discountBands } = extraCoverData.extraCover;

  if (itemValueAUD <= thresholdAUD) {
    return 0;
  }

  const discount = discountBands[discountBand] || 0;
  // Formula: (ItemValue - 100) / 100 × $4 × (1 - discount)
  const coverUnits = (itemValueAUD - thresholdAUD) / 100;
  const cost = coverUnits * basePricePer100 * (1 - discount);

  return Math.round(cost * 100) / 100;
}

/**
 * Calculate US tariff duties
 * @param {number} itemValueAUD - Item value in AUD
 * @param {string} countryOfOrigin - Country of manufacture
 * @returns {number} - Tariff amount in AUD
 */
function calculateTariffDuties(itemValueAUD, countryOfOrigin) {
  const tariffRate = getTariffRate(countryOfOrigin);
  return Math.round(itemValueAUD * tariffRate * 100) / 100;
}

/**
 * Calculate Zonos processing fees
 * @param {number} tariffAmount - The tariff duties amount in AUD
 * @returns {number} - Zonos fees in AUD
 */
function calculateZonosFees(tariffAmount) {
  const { processingChargePercent, flatFeeAUD } = tariffs.zonos;
  const percentageFee = tariffAmount * processingChargePercent;
  const total = percentageFee + flatFeeAUD;
  return Math.round(total * 100) / 100;
}

/**
 * Check if extra cover warning should be shown
 * @param {number} itemValueAUD - Item value in AUD
 * @param {boolean} hasExtraCover - Whether extra cover is enabled
 * @returns {boolean} - True if warning should be shown
 */
function shouldWarnExtraCover(itemValueAUD, hasExtraCover) {
  return itemValueAUD >= extraCoverData.extraCover.warningThresholdAUD && !hasExtraCover;
}

/**
 * Calculate total shipping cost to USA including all fees
 * @param {Object} params - Calculation parameters
 * @param {number} params.itemValueAUD - Item value in AUD
 * @param {string} params.weightBand - Weight category
 * @param {string} params.brandName - Brand name (for COO lookup)
 * @param {string} [params.countryOfOrigin] - Override COO (optional)
 * @param {boolean} [params.includeExtraCover=false] - Include insurance
 * @param {number} [params.discountBand=0] - AusPost discount band
 * @returns {Object} - Detailed breakdown of costs
 */
function calculateUSAShipping(params) {
  const {
    itemValueAUD,
    weightBand,
    brandName,
    countryOfOrigin: overrideCOO,
    includeExtraCover = false,
    discountBand = 0
  } = params;

  const zone = '3-USA & Canada';
  const countryOfOrigin = overrideCOO || getCountryOfOrigin(brandName);
  const tariffRate = getTariffRate(countryOfOrigin);

  // Calculate components
  const ausPostShipping = calculateAusPostShipping(zone, weightBand, discountBand);
  const extraCover = includeExtraCover ? calculateExtraCover(itemValueAUD, discountBand) : 0;
  const tariffDuties = calculateTariffDuties(itemValueAUD, countryOfOrigin);
  const zonosFees = calculateZonosFees(tariffDuties);

  const shippingSubtotal = ausPostShipping + extraCover;
  const dutiesSubtotal = tariffDuties + zonosFees;
  const totalShipping = shippingSubtotal + dutiesSubtotal;

  return {
    inputs: {
      itemValueAUD,
      weightBand,
      brandName,
      countryOfOrigin,
      tariffRate,
      includeExtraCover,
      discountBand
    },
    breakdown: {
      ausPostShipping,
      extraCover,
      shippingSubtotal,
      tariffDuties,
      zonosFees,
      dutiesSubtotal
    },
    totalShipping: Math.round(totalShipping * 100) / 100,
    warnings: {
      extraCoverRecommended: shouldWarnExtraCover(itemValueAUD, includeExtraCover)
    }
  };
}

/**
 * Get weight band from weight in grams
 * @param {number} weightGrams - Weight in grams
 * @returns {string} - Weight band name
 */
function getWeightBandFromGrams(weightGrams) {
  if (weightGrams < 250) return 'XSmall';
  if (weightGrams < 500) return 'Small';
  if (weightGrams < 1000) return 'Medium';
  if (weightGrams < 1500) return 'Large';
  return 'XLarge';
}

/**
 * Get all available brands
 * @returns {string[]} - Array of brand names
 */
function getAvailableBrands() {
  return Object.keys(brands.brands).sort();
}

/**
 * Get all available weight bands
 * @returns {Object[]} - Array of weight band info
 */
function getWeightBands() {
  const zone = postalRates.zones['3-USA & Canada'];
  return Object.entries(zone.weightBands).map(([key, data]) => ({
    key,
    label: data.label,
    maxWeight: data.maxWeight,
    basePrice: data.basePrice
  }));
}

/**
 * Get all countries with tariff rates
 * @returns {Object[]} - Array of country/rate pairs
 */
function getTariffCountries() {
  return Object.entries(tariffs.usaTariffs.rates).map(([country, rate]) => ({
    country,
    rate,
    ratePercent: Math.round(rate * 100)
  }));
}

module.exports = {
  calculateUSAShipping,
  calculateAusPostShipping,
  calculateExtraCover,
  calculateTariffDuties,
  calculateZonosFees,
  getCountryOfOrigin,
  getTariffRate,
  getWeightBandFromGrams,
  getAvailableBrands,
  getWeightBands,
  getTariffCountries,
  shouldWarnExtraCover
};

File: internal/calculator/calculator.go
package calculator

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// ShippingResult holds the complete calculation breakdown
type ShippingResult struct {
	Inputs    ShippingInputs    `json:"inputs"`
	Breakdown ShippingBreakdown `json:"breakdown"`
	Total     float64           `json:"totalShipping"`
	Warnings  ShippingWarnings  `json:"warnings"`
}

// ShippingInputs captures the input parameters
type ShippingInputs struct {
	ItemValueAUD      float64 `json:"itemValueAUD"`
	WeightBand        string  `json:"weightBand"`
	BrandName         string  `json:"brandName"`
	CountryOfOrigin   string  `json:"countryOfOrigin"`
	TariffRate        float64 `json:"tariffRate"`
	IncludeExtraCover bool    `json:"includeExtraCover"`
	DiscountBand      int     `json:"discountBand"`
}

// ShippingBreakdown shows individual cost components
type ShippingBreakdown struct {
	AusPostShipping  float64 `json:"ausPostShipping"`
	ExtraCover       float64 `json:"extraCover"`
	ShippingSubtotal float64 `json:"shippingSubtotal"`
	TariffDuties     float64 `json:"tariffDuties"`
	ZonosFees        float64 `json:"zonosFees"`
	DutiesSubtotal   float64 `json:"dutiesSubtotal"`
}

// ShippingWarnings holds any warnings for the user
type ShippingWarnings struct {
	ExtraCoverRecommended bool `json:"extraCoverRecommended"`
}

// GetCountryOfOrigin returns the COO for a brand, or default
func GetCountryOfOrigin(brandName string) string {
	if brand, ok := Brands[brandName]; ok {
		return brand.PrimaryCOO
	}
	return DefaultCOO
}

// GetTariffRate returns the US tariff rate for a country
func GetTariffRate(country string) float64 {
	if rate, ok := USATariffs.Rates[country]; ok {
		return rate
	}
	return USATariffs.Rates[DefaultCOO]
}

// CalculateAusPostShipping calculates the AusPost shipping cost
func CalculateAusPostShipping(zone, weightBand string, discountBand int) (float64, error) {
	zoneData, ok := PostalZones[zone]
	if !ok {
		return 0, fmt.Errorf("unknown zone: %s", zone)
	}

	weightData, ok := zoneData.WeightBands[weightBand]
	if !ok {
		return 0, fmt.Errorf("unknown weight band: %s", weightBand)
	}

	discount, ok := zoneData.DiscountBands[discountBand]
	if !ok {
		discount = 0
	}

	// Formula: Base × (1 + handling) × (1 - discount)
	withHandling := weightData.BasePrice * (1 + zoneData.HandlingFee)
	finalPrice := withHandling * (1 - discount)

	return round2(finalPrice), nil
}

// CalculateExtraCover calculates insurance cost
func CalculateExtraCover(itemValueAUD float64, discountBand int) float64 {
	if itemValueAUD <= ExtraCover.ThresholdAUD {
		return 0
	}

	discount, ok := ExtraCover.DiscountBands[discountBand]
	if !ok {
		discount = 0
	}

	// Formula: (ItemValue - 100) / 100 × $4 × (1 - discount)
	coverUnits := (itemValueAUD - ExtraCover.ThresholdAUD) / 100
	cost := coverUnits * ExtraCover.BasePricePer100 * (1 - discount)

	return round2(cost)
}

// CalculateTariffDuties calculates US import duties
func CalculateTariffDuties(itemValueAUD float64, countryOfOrigin string) float64 {
	rate := GetTariffRate(countryOfOrigin)
	return round2(itemValueAUD * rate)
}

// CalculateZonosFees calculates Zonos processing fees
func CalculateZonosFees(tariffAmount float64) float64 {
	percentageFee := tariffAmount * Zonos.ProcessingChargePercent
	total := percentageFee + Zonos.FlatFeeAUD
	return round2(total)
}

// ShouldWarnExtraCover returns true if extra cover warning should show
func ShouldWarnExtraCover(itemValueAUD float64, hasExtraCover bool) bool {
	return itemValueAUD >= ExtraCover.WarningThresholdAUD && !hasExtraCover
}

// CalculateUSAShippingParams holds parameters for the main calculation
type CalculateUSAShippingParams struct {
	ItemValueAUD      float64
	WeightBand        string
	BrandName         string
	CountryOfOrigin   string // optional override
	IncludeExtraCover bool
	DiscountBand      int
}

// CalculateUSAShipping performs the complete shipping calculation
func CalculateUSAShipping(params CalculateUSAShippingParams) (*ShippingResult, error) {
	zone := "3-USA & Canada"

	// Determine country of origin
	coo := params.CountryOfOrigin
	if coo == "" {
		coo = GetCountryOfOrigin(params.BrandName)
	}
	tariffRate := GetTariffRate(coo)

	// Calculate components
	ausPostShipping, err := CalculateAusPostShipping(zone, params.WeightBand, params.DiscountBand)
	if err != nil {
		return nil, err
	}

	var extraCover float64
	if params.IncludeExtraCover {
		extraCover = CalculateExtraCover(params.ItemValueAUD, params.DiscountBand)
	}

	tariffDuties := CalculateTariffDuties(params.ItemValueAUD, coo)
	zonosFees := CalculateZonosFees(tariffDuties)

	shippingSubtotal := ausPostShipping + extraCover
	dutiesSubtotal := tariffDuties + zonosFees
	total := shippingSubtotal + dutiesSubtotal

	return &ShippingResult{
		Inputs: ShippingInputs{
			ItemValueAUD:      params.ItemValueAUD,
			WeightBand:        params.WeightBand,
			BrandName:         params.BrandName,
			CountryOfOrigin:   coo,
			TariffRate:        tariffRate,
			IncludeExtraCover: params.IncludeExtraCover,
			DiscountBand:      params.DiscountBand,
		},
		Breakdown: ShippingBreakdown{
			AusPostShipping:  ausPostShipping,
			ExtraCover:       extraCover,
			ShippingSubtotal: shippingSubtotal,
			TariffDuties:     tariffDuties,
			ZonosFees:        zonosFees,
			DutiesSubtotal:   dutiesSubtotal,
		},
		Total: round2(total),
		Warnings: ShippingWarnings{
			ExtraCoverRecommended: ShouldWarnExtraCover(params.ItemValueAUD, params.IncludeExtraCover),
		},
	}, nil
}

// GetWeightBandFromGrams returns the weight band for a given weight
func GetWeightBandFromGrams(weightGrams int) string {
	switch {
	case weightGrams < 250:
		return "XSmall"
	case weightGrams < 500:
		return "Small"
	case weightGrams < 1000:
		return "Medium"
	case weightGrams < 1500:
		return "Large"
	default:
		return "XLarge"
	}
}

// GetAvailableBrands returns all brand names sorted
func GetAvailableBrands() []string {
	brands := make([]string, 0, len(Brands))
	for name := range Brands {
		brands = append(brands, name)
	}
	sort.Strings(brands)
	return brands
}

// WeightBandInfo holds weight band details for API responses
type WeightBandInfo struct {
	Key       string  `json:"key"`
	Label     string  `json:"label"`
	MaxWeight int     `json:"maxWeight"`
	BasePrice float64 `json:"basePrice"`
}

// GetWeightBands returns all weight bands for USA zone
func GetWeightBands() []WeightBandInfo {
	zone := PostalZones["3-USA & Canada"]
	bands := make([]WeightBandInfo, 0, len(zone.WeightBands))

	// Order matters for display
	order := []string{"XSmall", "Small", "Medium", "Large", "XLarge"}
	for _, key := range order {
		if wb, ok := zone.WeightBands[key]; ok {
			bands = append(bands, WeightBandInfo{
				Key:       key,
				Label:     wb.Label,
				MaxWeight: wb.MaxWeight,
				BasePrice: wb.BasePrice,
			})
		}
	}
	return bands
}

// TariffCountryInfo holds tariff info for API responses
type TariffCountryInfo struct {
	Country     string  `json:"country"`
	Rate        float64 `json:"rate"`
	RatePercent int     `json:"ratePercent"`
}

// GetTariffCountries returns all countries with tariff rates
func GetTariffCountries() []TariffCountryInfo {
	countries := make([]TariffCountryInfo, 0, len(USATariffs.Rates))
	for country, rate := range USATariffs.Rates {
		countries = append(countries, TariffCountryInfo{
			Country:     country,
			Rate:        rate,
			RatePercent: int(math.Round(rate * 100)),
		})
	}
	sort.Slice(countries, func(i, j int) bool {
		return countries[i].Country < countries[j].Country
	})
	return countries
}

// round2 rounds to 2 decimal places
func round2(val float64) float64 {
	return math.Round(val*100) / 100
}

// ZoneShippingResult holds calculation results for a single zone
type ZoneShippingResult struct {
	ZoneID      string            `json:"zoneId"`      // e.g., "1-New Zealand"
	ZoneName    string            `json:"zoneName"`    // e.g., "New Zealand"
	Inputs      ShippingInputs    `json:"inputs"`
	Breakdown   ShippingBreakdown `json:"breakdown"`
	Total       float64           `json:"totalShipping"`
	Warnings    ShippingWarnings  `json:"warnings"`
	HasTariffs  bool              `json:"hasTariffs"`  // Whether this zone applies tariffs
}

// MultiZoneResult holds calculation results for all zones
type MultiZoneResult struct {
	Zones []ZoneShippingResult `json:"zones"`
}

// CalculateAllZonesParams holds parameters for multi-zone calculation
type CalculateAllZonesParams struct {
	ItemValueAUD      float64
	WeightBand        string
	BrandName         string
	CountryOfOrigin   string // optional override
	IncludeExtraCover bool
	DiscountBand      int
}

// CalculateAllZones performs shipping calculation for all zones
func CalculateAllZones(params CalculateAllZonesParams) (*MultiZoneResult, error) {
	// Determine country of origin
	coo := params.CountryOfOrigin
	if coo == "" {
		coo = GetCountryOfOrigin(params.BrandName)
	}

	// Get all zones in a consistent order
	zoneOrder := []string{"1-New Zealand", "3-USA & Canada", "4-UK & Ireland"}
	results := make([]ZoneShippingResult, 0, len(zoneOrder))

	for _, zoneID := range zoneOrder {
		_, ok := PostalZones[zoneID]
		if !ok {
			continue // Skip if zone not found
		}

		// Determine if this zone has tariffs (only USA)
		hasTariffs := zoneID == "3-USA & Canada"

		// Calculate components
		ausPostShipping, err := CalculateAusPostShipping(zoneID, params.WeightBand, params.DiscountBand)
		if err != nil {
			return nil, fmt.Errorf("zone %s: %w", zoneID, err)
		}

		var extraCover float64
		if params.IncludeExtraCover {
			extraCover = CalculateExtraCover(params.ItemValueAUD, params.DiscountBand)
		}

		shippingSubtotal := ausPostShipping + extraCover

		// Calculate tariffs and duties (only for USA)
		var tariffDuties, zonosFees, dutiesSubtotal float64
		var tariffRate float64
		if hasTariffs {
			tariffRate = GetTariffRate(coo)
			tariffDuties = CalculateTariffDuties(params.ItemValueAUD, coo)
			zonosFees = CalculateZonosFees(tariffDuties)
			dutiesSubtotal = tariffDuties + zonosFees
		}

		total := shippingSubtotal + dutiesSubtotal

		// Extract zone name from zone ID (e.g., "1-New Zealand" -> "New Zealand")
		zoneName := zoneID
		if idx := strings.Index(zoneID, "-"); idx >= 0 && idx < len(zoneID)-1 {
			zoneName = zoneID[idx+1:]
		}

		results = append(results, ZoneShippingResult{
			ZoneID:   zoneID,
			ZoneName: zoneName,
			Inputs: ShippingInputs{
				ItemValueAUD:      params.ItemValueAUD,
				WeightBand:        params.WeightBand,
				BrandName:         params.BrandName,
				CountryOfOrigin:   coo,
				TariffRate:        tariffRate,
				IncludeExtraCover: params.IncludeExtraCover,
				DiscountBand:      params.DiscountBand,
			},
			Breakdown: ShippingBreakdown{
				AusPostShipping:  ausPostShipping,
				ExtraCover:       extraCover,
				ShippingSubtotal: shippingSubtotal,
				TariffDuties:     tariffDuties,
				ZonosFees:        zonosFees,
				DutiesSubtotal:   dutiesSubtotal,
			},
			Total: round2(total),
			Warnings: ShippingWarnings{
				ExtraCoverRecommended: ShouldWarnExtraCover(params.ItemValueAUD, params.IncludeExtraCover),
			},
			HasTariffs: hasTariffs,
		})
	}

	return &MultiZoneResult{
		Zones: results,
	}, nil
}
